README.md
## EINSim: Statistically Inferring ECC Characteristics Using Only Post-Correction Errors 

The Error Inference Simulator (EINSim) uses Monte-Carlo simulation to model how different error-correction code (ECC) schemes transform the spatial distribution of errors in a memory device (e.g., DRAM data-retention errors). EINSim simulates an ECC dataword's lifetime throughout ECC encoding, error injection, and ECC decoding. By simulating enough datawords (e.g., > 1e6), the statistical distribution of post-correction errors approximates that of a real device. 

EINSim is a key part of the EIN statistical inference methodology originally proposed in our DSN 2019 academic paper [1]. EIN uses maximum-a-posteriori (MAP) estimation over different ECC models (computed using EINSim) to infer a device's pre-correction error characteristics from post-correction error distributions. This information can be used to:

- Reverse-engineer characteristics of an unknown ECC scheme (e.g., ECC data-/code-word length, ECC correction capability)
- Infer pre-correction error rates that are obfuscated by ECC

EIN is especially useful for devices in which the pre-correction error characteristics are hidden (e.g., DRAM with on-die ECC). Our paper [1] explains EIN and its potential applications in much greater detail.

*Please send questions to Minesh Patel at minesh.patelh@gmail.com*

## Codebase Overview

Current version: 2.0.0

EINSim comprises two parts:
- A C++ command-line tool that simulates the raw statistical distributions
- Python scripts that apply the simulation results to real experimental data

We use Doxygen to document the source code and provide a Doxyfile for building HTML and LaTeX documentation. To build the documentation, simply issue:

```
$ doxygen
```
when in the project directory, or point ```doxygen``` to the provided Doxyfile. The HTML documentation will be built under ```doxygen/html/index.html```. 

## Dependencies

Building and running EINSim requires the following working installations:

- A C++11 toolchain (e.g., GCC, Clang, MSVC)
	- pthreads

- Python 2/3
	- Matplotlib 2.1.1+ (for graph generation only)
	- SciPy
	- NumPy

EINSim has been tested with:

	- GCC 9.1.0
	- GCC 8.3.0
	- GCC 7.4.0
	- GCC 6.3.0
	- Apple LLVM 10.0.0 (clang-1000.11.45.5)
	- Apple LLVM 9.1.0 (clang-902.0.39.2)
	- Python 3.7.3 (Anaconda 3)
	- Python 3.7.2
	- Python 3.7.1
	- Python 3.5.3
	- Python 2.7.13

## Building in Linux

```
$ make [-j <# threads>] [other make options] <target>
```

The makefile has various targets, described as follows:

- ```release``` builds ```einsim``` with full optimizations
- ```debug``` builds ```einsim.d``` with no optimization and debug symbols
- ```lib``` builds ```einsim.so``` as a shared library with full optimizations
- ```doc``` builds Doxygen documentation in the directory ```doxygen```
- ```all``` builds both ```release``` and ```debug```
- ```clean``` cleans build and binary files for both ```release``` and ```debug```

Omitting the ```target``` argument defaults to the ```release``` configuration. 

## Usage

EINSim runs as a command-line tool with several CLI options that are shown when running EINSim without options:

```
$ ./path/to/einsim
```

Quickly testing whether EINSim works correctly can be done by issuing:
```
$ ./path/to/einsim -m t -T fast [-t <# threads>]
```
This will run a brief test of all implemented ECC schemes across various code parameters and will report success or failure. 

EINSim has three modes of operation:

- ```Simulation``` (```./path/to/einsim -m s```) is the primary mode of operation and simulates how the spatial distribution of errors is transformed by different ECC schemes.

- ```Test``` (```./path/to/einsim -m t```) is used primarily for sanity-checking and regression testing. It sweeps through a set of pre-baked configurations for each ECC scheme that are designed to roughly cover the range of important parameters for the scheme. This helps check for obvious errors when modifying, updating, or adding ECC schemes.  

- ```Debug``` (```./path/to/einsim -m d```) is used for debugging various components of EINSim. It is similar to the *simulation* mode, but provides a simpler core simulation loop with more controls over an ECC word. While *debug* mode is designed to make working with EINSim easier, it need not be used if the user is comfortable with a different debugging approach.

## EINSim high-level overview

Our paper [1] provides a detailed overview of EINSim's purpose and design, and we highly recommend the interested user to refer to it when delving deeply into EINSim.

At a high level, EINSim simulates the error characteristics of configurable-length *words*, which are referred to as *bursts* in the context of DRAM. Bursts are typically several bytes long (e.g., 32B) and may comprise multiple ECC data words. Therefore, EINSim subdivides each burst into one or more ECC datawords as per the particular ECC schemes that have been chosen for simulation. 

In the core simulation loop, EINSim repeatedly simulates individual bursts through the following steps:

1) Writing the burst with a configurable data pattern (e.g., 0xFF, RANDOM, custom data)
2) ECC encoding
3) Error injection according to error-mechanism parameters (e.g., error distribution, DRAM true-/anti-cell layout, data pattern)
4) ECC decoding
5) Measuring some property of the output (e.g., counting the number of errors in the decoded burst with respect to the originally written data pattern)

EINSim then outputs the final measured property using the output format described in the following section.

## JSON representation of ECC schemes

EINSim represents ECC schemes using JSON with the following members:

- **s**: type of ECC scheme (e.g., HSC for Hamming single-error correcting)
- **k**: dataword length (bits)
- **p**: random seed used to initially generate the ECC code
- **uid**: 64-bit unsigned identifier for this ECC scheme
- **G**: generator matrix
- **H**: parity-check matrix
- **R**: degenerator matrix
- **R**: degenerator matrix
- **miscorrection_profile**: miscorrection profile as used by [BEER](https://github.com/CMU-SAFARI/BEER) (ignored by EINSim)

As an example, a (7, 4, 3) Hamming code:

```
{"s":"HSC","k":4,"p":0,"uid":3285674560482760615,"G":[[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1],[1,1,0,1],[0,1,1,1],[1,0,1,1]],"H":[[1,1,0,1,1,0,0],[0,1,1,1,0,1,0],[1,0,1,1,0,0,1]],"R":[[1,0,0,0,0,0,0],[0,1,0,0,0,0,0],[0,0,1,0,0,0,0],[0,0,0,1,0,0,0]]}
```

Additional examples are provided under the ```examples``` directory.

## EINSim output format

The EINSim output format is verbose by design. Each line of output contains the pre- and post-correction error distribution when simulating N words for a single system configuration. This keeps it relatively easy to search for data concerning a particular configuration using common tools such as ```awk``` or ```grep```.

Since a large N (e.g., N > 1e6) is required to achieve a statistically representative sample of the true error distributions (discussed further in Sections 4 and 5 of Patel et al. [1]), EINSim is designed for high parallelization. EINSim typically divides N into smaller units, and each unit outputs its results as soon as it finishes. The outputs across all units can later be combined into a single distribution using the provided Python script ```script/utils/aggregate_output.py```.

The output format of a single output line is described as follows and provides reasonable human-readability:

```
[DATA] <ECC scheme>: p:<int> t:<int> k:<int> n:<int> m:<int> rber:<float> bl:<int> bcl:<int> ps:<int> ed:<string> cd:<string> dp:<string> [ <error count>:<# words pre-correction>:<# words post-correction> <>:<>:<> ... repeat ]
```

We briefly explain each of the fields:

- **uid**: 64-bit unsigned UID computed for the ECC scheme based on its generator, parity-check, and degenerator matrices - used to identify the ECC scheme
- **nw (number of words)**: number of ECC words simulated within this line of output
- **bl (burst length)**: size of the simulated burst (data bits only; will be subdivided into ECC datawords as required)
- **bcl (burst codeword length)**: size of the simulated burst, including ECC check bits
- **ps (pad size)**: additional zero-padding used for when a burst unevenly divides into ECC words
- **em (error model)**: error model used for error-injection, including parameter values
- **cd (true-/anti-cell distribution)**: the spatial distribution of true-/anti-cells to assume when simulating bursts
- **dp (data pattern)**: the data pattern to assume for all datawords (e.g., 0xFF, RANDOM) 
- **obs (observable)**: the metric to compute across all simulated ECC words (e.g., per-bit error counts)
- **[data]**: data values corresponding to the requested observable

Example output for a BCH 3EC code with a 511-bit codeword, 484-bit dataword, and a 256-bit burst length with a UNIFORM_RANDOM pre-correction error distribution, an equal number of true-/anti-cells, and a RANDOM data pattern:

```
$ ./path/to/einsim -m s -s HSC -k 4 -b 4 -p 0 -w BLOCKS -c ALL_TRUE -n 100 -d ALL_ONES -o PER_BIT_ERROR_COUNT -e DATA_RETENTION,0.5
...
[DATA] uid:3285674560482760615 nw:100 bl:4 bcl:7 ps:0 em:DATA_RETENTION(p:0.500000) cd:ALL_TRUE dp:ALL_ONES obs:PER_BIT_ERROR_COUNT [ 52 47 42 52 : 56 45 42 53 52 53 45 ]
```

The script ```script/aggregate_output.py``` can be used to aggregate multiple outputs across multiple runs (or simply within just one run) as follows:

```
$ ./path/to/script/aggregate_output.py -o <filename: destination> [<filename: input EINSim results 0> <filename: input EINSim results 1> ...]
```

Note: it is *not* always necessary to aggregate simulation outputs, but aggregation saves considerable computation time and disk space when applying many simulation results to real experimental data.

## Applying real experimental data to EINSim output

We provide a Python script ```script/ein/ein.py``` to apply EINSim results to experimental data. ```ein.py``` can be run as follows:

```
$ ./path/to/script/ein/ein.py <simulation data filename> [<simulation data filename> ...] -e <experimental data filename>
```

Experimental data files use same format as simulation outputs for convenience. However, most of the parameters are ignored and may be set to **-1**.

## Using the provided examples

Under ```example```, we provide several example configuration files for both ECC schemes and error models. These can be used as inputs to EINSim as alternatives to passing command-line arguments and are intended to simplify the use of complicated configurations and uses that include automation.

## Running the examples given in the original publication [1]

EINSim tag "full-size-dataset-v1.0" provides full working examples of running EINSim end-to-end, including the *simulation* data used in our paper [1]. Because this EINSim release is not fully backwards-compatible with the initial release, please refer to the README within tag "full-size-dataset-v1.0" for instructions on how to run the original version.

## Licensing

The current version of the simulator is provided as-is under the MIT license.

The following libraries are used and are located under ```lib``` with their own license:
- [Eigen](http://eigen.tuxfamily.org/index.php?title=Main_Page)
- [cxxopts](https://github.com/jarro2783/cxxopts)
- libtp - Inspired by [CTPL](https://github.com/vit-vit/CTPL)
- [rapidjson](https://rapidjson.org/)
- crc64 - CRC64 routines written by Salvatore Sanfilippo

The BCH encoding/decoding helper functions are found under src/codes/bch_helpers* and are adapted from bch3.c found at http://eccpage.com/. The original license provided by the author, Robert Morelos-Zaragoza, is retained along with a brief description of the modifications we have made.

## Attribution

Please cite the following paper when using EINSim:

[\[1\] Minesh Patel, Jeremie S. Kim, Hasan Hassan, and Onur Mutlu, "Understanding and Modeling On-Die Error Correction in Modern DRAM: An Experimental Study Using Real Devices", in the Proceedings of the 49th Annual IEEE/IFIP International Conference on Dependable Systems and Networks (DSN 2019), Portland, OR, USA, June 2019.](https://people.inf.ethz.ch/omutlu/pub/understanding-and-modeling-in-DRAM-ECC_dsn19.pdf)


CHANGES
version 1.0.0 [2019 May 10]
	+ The first public release of EINSim

version 2.0.0 [2020 Oct 12]
    + Addding the following major features:
        + Support for multiple computing multiple different observations over the final simulated ECC words
        + JSON representation of ECC schemes
        + Flexible error-injection models per-bit (+ JSON configuration files)
        + Custom data patterns
        + UID identification for ECC scheme configurations
        + Configurable word-to-burst mappings
    + Improving overall code style (e.g., better use of `const' qualifiers)
    + Updating examples and scripts to support the new features
    + Updating documentation with theses changes

version 2.1.0 [2023 Dec 7]
    + Adding a shared library build target for EINSim


script/utils.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# @file utils.py
#
# @brief Common helper functions
#
# @author Minesh Patel
# Contact: minesh.patelh@gmail.com
import sys
import os
import traceback
try:
    import lzma
except ImportError:
    print("[WARNING] unable to import lzma module required to read lzma/xz files. Use Python3.3+ or backports lzma for this functionality")
import gzip
import bz2

def add_to_hist(base, addend):
    for before, after, count in addend:
        if before not in base:
            base[before] = {}
        if after not in base[before]:
            base[before][after] = 0
        base[before][after] += count

def open_sim_file(fname):
    def warn_about_compressed_progress_bar():
        print("\t[WARNING] progress bar will go >100% for compressed file. On the bright side, the higher it goes, the better your compression ratio is :D")
    
    if fname[-3:] == '.xz':
        warn_about_compressed_progress_bar()
        return lzma.open(fname, mode='rt')
    elif fname[-3:] == '.gz':
        warn_about_compressed_progress_bar()
        return gzip.open(fname, mode='rt')
    elif fname[-4:] == '.bz2':
        warn_about_compressed_progress_bar()
        return bz2.open(fname, mode='rt')
    else:
        return open(fname, 'r')

# example input lines
# [DATA] REP: p:584576191 t:1 k:4 n:12 rber:1e-06 bl:256 bcl:768 ps:0 ed:UNIFORM_RANDOM cd:ALL_TRUE_OR_ALL_ANTI dp:RANDOM [ 0:9998:10000 1:2:0 ]
# [DATA] REP: p:584576191 t:3 k:4 n:28 rber:0.0331251 bl:256 bcl:1792 ps:0 ed:UNIFORM_RANDOM cd:ALL_TRUE_OR_ALL_ANTI dp:RANDOM [ 0:0:9051 1:0:900 2:0:49 31:2:0 32:2:0 33:1:0 34:3:0 35:4:0 36:9:0 37:16:0 38:16:0 39:24:0 40:34:0 41:44:0 42:70:0 43:72:0 44:98:0 45:131:0 46:135:0 47:182:0 48:190:0 49:234:0 50:252:0 51:309:0 52:276:0 53:365:0 54:407:0 55:413:0 56:426:0 57:437:0 58:449:0 59:431:0 60:499:0 61:458:0 62:426:0 63:357:0 64:402:0 65:353:0 66:365:0 67:303:0 68:298:0 69:252:0 70:203:0 71:173:0 72:150:0 73:146:0 74:114:0 75:98:0 76:79:0 77:59:0 78:55:0 79:44:0 80:25:0 81:31:0 82:16:0 83:18:0 84:15:0 85:7:0 86:4:0 87:5:0 88:2:0 89:3:0 92:4:0 93:1:0 95:1:0 99:1:0 101:1:0 ]
def parse_einsim_output_line(line):
    if len(line) > 6 and line[:6] == "[DATA]" in line:
        meta, raw_data = line[7:].strip().split('[')
        meta_split = meta.split(' ')
        ecc_category = meta_split[0][:-1]
        if ecc_category == 'REP':
            p = int(meta_split[1][2:])
            t = int(meta_split[2][2:])
            k = int(meta_split[3][2:])
            n = int(meta_split[4][2:])
            m = -1
            rber = float(meta_split[5][5:])
            bl = int(meta_split[6][3:])
            bcl = int(meta_split[7][4:])
            ps = int(meta_split[8][3:])
            ed = meta_split[9][3:]
            cd = meta_split[10][3:]
            dp = meta_split[11][3:]
        elif ecc_category == 'HSC':
            p = int(meta_split[1][2:])
            t = int(meta_split[2][2:])
            k = int(meta_split[3][2:])
            n = int(meta_split[4][2:])
            m = -1
            rber = float(meta_split[5][5:])
            bl = int(meta_split[6][3:])
            bcl = int(meta_split[7][4:])
            ps = int(meta_split[8][3:])
            ed = meta_split[9][3:]
            cd = meta_split[10][3:]
            dp = meta_split[11][3:]
        elif ecc_category == 'BCH':
            p = int(meta_split[1][2:])
            t = int(meta_split[2][2:])
            k = int(meta_split[3][2:])
            n = int(meta_split[4][2:])
            m = int(meta_split[5][2:])
            rber = float(meta_split[6][5:])
            bl = int(meta_split[7][3:])
            bcl = int(meta_split[8][4:])
            ps = int(meta_split[9][3:])
            ed = meta_split[10][3:]
            cd = meta_split[11][3:]
            dp = meta_split[12][3:]
        elif ecc_category == 'UNK':
            p = int(meta_split[1][2:])
            t = int(meta_split[2][2:])
            k = int(meta_split[3][2:])
            n = int(meta_split[4][2:])
            # only BCH has the extra m param
            if meta_split[5][0] == 'm':
                m = int(meta_split[5][2:])
                off = 1
            else:
                m = -1
                off = 0
            rber = float(meta_split[5 + off][5:])
            bl = int(meta_split[6 + off][3:])
            bcl = int(meta_split[7 + off][4:])
            ps = int(meta_split[8 + off][3:])
            ed = meta_split[9 + off][3:]
            cd = meta_split[10 + off][3:]
            dp = meta_split[11 + off][3:]
        else:
            print("[WARNING] unhandled ECC category \"" + ecc_category + "\" at line: " + str(line_num) + " in file " + infilename)
            traceback.print_exc()
            return None
        
        metadata = [ecc_category, p, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp]
        data = [map(int, entry.split(':')) for entry in raw_data.split(' ') if ':' in entry]

        # print(ecc_type, permutation, nerrs_correctable, n_data_bits, n_code_bits, dp, error_model)
        # print(hist_data)
        return [metadata, data]
    else:
        return None

def parse_einsim_output_file(infilename):
    dirty_data = []
    malformed_line = -1
    file_size_bytes = os.path.getsize(infilename)    
    with open_sim_file(infilename) as f:
        line_num = 0
        try:
            line = f.readline()
            while line:
                line_num += 1
                try:
                    parsed_line = parse_einsim_output_line(line)
                    if parsed_line:
                        dirty_data.append(parsed_line)
                except KeyboardInterrupt:
                    raise
                except:
                    if malformed_line == -1:
                        print("[WARNING] malformed line in file " + infilename + " at line: " + str(line_num) + ' - ', sys.exc_info()[0])
                        print("    bad line:", line)
                        traceback.print_exc()
                        malformed_line = line_num
                    pass

                if line_num == 1 or (line_num % 1000) == 0:
                    sys.stdout.write('\t[' + infilename + "] {0:.2f}".format(100.0 * f.tell() / float(file_size_bytes)) + '% complete  \r')
                line = f.readline()
        except:
            print("Unexpected error in file " + infilename + " at line " + str(line_num) + ':', sys.exc_info()[0])
            raise

    return dirty_data

def run_uid_from_metadata(metadata):
    ecc_category, p, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp = metadata

    run_uid = '[DATA]'
    if ecc_category == 'REP':
        run_uid += ' REP:'
        run_uid += ' p:' + '-1'
        run_uid += ' t:' + str(t)
        run_uid += ' k:' + str(k)
        run_uid += ' n:' + str(n)
        run_uid += ' rber:' + str(rber)
        run_uid += ' bl:' + str(bl)
        run_uid += ' bcl:' + str(bcl)
        run_uid += ' ps:' + str(ps)
        run_uid += ' ed:' + str(ed)
        run_uid += ' cd:' + str(cd)
        run_uid += ' dp:' + str(dp)
    elif ecc_category == 'HSC':
        run_uid += ' HSC:'
        run_uid += ' p:' + '-1'
        run_uid += ' t:' + str(t)
        run_uid += ' k:' + str(k)
        run_uid += ' n:' + str(n)
        run_uid += ' rber:' + str(rber)
        run_uid += ' bl:' + str(bl)
        run_uid += ' bcl:' + str(bcl)
        run_uid += ' ps:' + str(ps)
        run_uid += ' ed:' + str(ed)
        run_uid += ' cd:' + str(cd)
        run_uid += ' dp:' + str(dp)
    elif ecc_category == 'BCH':
        run_uid += ' BCH:'
        run_uid += ' p:' + '-1'
        run_uid += ' t:' + str(t)
        run_uid += ' k:' + str(k)
        run_uid += ' n:' + str(n)
        run_uid += ' m:' + str(m)
        run_uid += ' rber:' + str(rber)
        run_uid += ' bl:' + str(bl)
        run_uid += ' bcl:' + str(bcl)
        run_uid += ' ps:' + str(ps)
        run_uid += ' ed:' + str(ed)
        run_uid += ' cd:' + str(cd)
        run_uid += ' dp:' + str(dp)
    elif ecc_category == 'UNK':
        run_uid += ' UNK:'
        run_uid += ' bl:' + str(bl)
    else:
        print("[WARNING] unhandled ECC category \"" + ecc_category)

    return run_uid

def parse_all_files(infilenames, experimental=False):
    raw_data = {}
    experiment_id = 0

    # extract data from all files and aggregrate it
    print("parsing %d files" % len(infilenames))
    for filename in infilenames:
        dirty_data = parse_einsim_output_file(filename)
        # print(dirty_data)
        for metadata, data in dirty_data:
            ecc_category, p, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp = metadata
            
            # ensure that experiments get unique UIDs - we do not aggregate them
            run_uid = run_uid_from_metadata(metadata)
            if experimental:
                run_uid += '_experiment_' + str(experiment_id)
                experiment_id += 1

            # omit the permutation - we're aggregating over it!
            all_metadata = [ecc_category, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp]
            if run_uid not in raw_data:
                raw_data[run_uid] = {'metadata' : all_metadata, 'data' : [], 'p' : []}
            raw_data[run_uid]['data'].append(data)
            raw_data[run_uid]['p'].append(p)

    # print("Done parsing all files. Got:", len(raw_data), "files worth")

    # combine histograms
    for run_uid, components in raw_data.items():
        ecc_category, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp = components['metadata']
        
        raw_histogram = {}
        re_histogram = {}
        ue_histogram = {}
        for run in components['data']:
            for n_errs, re_count, ue_count in run:
                if n_errs not in raw_histogram:
                    raw_histogram[n_errs] = [0, 0]
                raw_histogram[n_errs][0] += re_count
                raw_histogram[n_errs][1] += ue_count

                if re_count > 0:
                    assert n_errs <= bcl, "Cannot have more errors in a burst_codeword than there are bits: errs:" + str(n_errs) + ", bcl:" + str(bcl)
                    if n_errs not in re_histogram:
                        re_histogram[n_errs] = 0
                    re_histogram[n_errs] += re_count
                if ue_count > 0:
                    assert n_errs <= bl, "Cannot have more errors in a burst than there are bits: errs:" + str(n_errs) + ", bl:" + str(bl)
                    if n_errs not in ue_histogram:
                        ue_histogram[n_errs] = 0
                    ue_histogram[n_errs] += ue_count

        # compute the RBER and UBER
        if re_histogram:
            # account for the pad size in the RBER - a real device wouldn't actually use those bits!
            rber_calculated = sum([n_errs * count for n_errs, count in re_histogram.items()]) / float((bcl - ps) * sum(re_histogram.values()))
        else:
            rber_calculated = -1.0
        uber_calculated = sum([n_errs * count for n_errs, count in ue_histogram.items()]) / float(bl * sum(ue_histogram.values()))
        raw_data[run_uid] = {'metadata' : components['metadata']
                            , 'p' : components['p']
                            , 'raw_data' : raw_histogram
                            , 'rber' : rber_calculated
                            , 'uber' : uber_calculated
                            , 're_data' : re_histogram
                            , 'ue_data': ue_histogram}

    return raw_data

def get_minimized_output_string(raw_data):
    out_str = ""
    for key, val in raw_data.items(): 
        out_str += key
        out_str += ' [ '
        out_str += ' '.join([':'.join(map(str, [pair[0], pair[1][0], pair[1][1]])) for pair in val['raw_data'].items()])
        out_str += ' ]\n'
    return out_str
    
# callback(run_uid, run_data, *callback_args) is called for every line parsed
# assumes that the input has already been aggregated! if not, repeat models will be treated as independent models
def parse_files_incremental(infilenames, is_experimental, callback, *callback_args):
    experiment_id = 0

    # extract data from all files and aggregrate it
    print("parsing %d files" % len(infilenames))
    for infilename in infilenames:

        # for each line in the file, invoke the callback
        malformed_line = -1
        file_size_bytes = os.path.getsize(infilename)
        with open_sim_file(infilename) as f:
            line_num = 0

            line = f.readline()
            while line:
                line_num += 1
                try:
                    parsed_line = parse_einsim_output_line(line)
                except KeyboardInterrupt:
                    raise
                except:
                    parsed_line = None
                    if malformed_line == -1:
                        print("[WARNING] malformed line in file " + infilename + " at line: " + str(line_num) + ' - ', sys.exc_info()[0])
                        print("    bad line:", line)
                        traceback.print_exc()
                        malformed_line = line_num
                    pass

                if parsed_line:
                    metadata, data = parsed_line
                    ecc_category, p, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp = metadata
                    
                    # ensure that experiments get unique UIDs - we do not aggregate them
                    run_uid = run_uid_from_metadata(metadata)
                    if is_experimental:
                        run_uid += '_experiment_' + str(experiment_id)
                        experiment_id += 1

                    # omit the permutation - we're aggregating over it!
                    metadata_no_p = [ecc_category, t, k, n, m, rber, bl, bcl, ps, ed, cd, dp]

                    raw_histogram = {}
                    re_histogram = {}
                    ue_histogram = {}
                
                    for n_errs, re_count, ue_count in data:
                        if n_errs not in raw_histogram:
                            raw_histogram[n_errs] = [0, 0]
                        raw_histogram[n_errs][0] += re_count
                        raw_histogram[n_errs][1] += ue_count

                        if re_count > 0:
                            assert n_errs <= bcl, "Cannot have more errors in a burst_codeword than there are bits: errs:" + str(n_errs) + ", bcl:" + str(bcl)
                            if n_errs not in re_histogram:
                                re_histogram[n_errs] = 0
                            re_histogram[n_errs] += re_count
                        if ue_count > 0:
                            assert n_errs <= bl, "Cannot have more errors in a burst than there are bits: errs:" + str(n_errs) + ", bl:" + str(bl)
                            if n_errs not in ue_histogram:
                                ue_histogram[n_errs] = 0
                            ue_histogram[n_errs] += ue_count

                    # compute the RBER and UBER
                    if re_histogram:
                        # account for the pad size in the RBER - a real device wouldn't actually use those bits!
                        rber_calculated = sum([n_errs * count for n_errs, count in re_histogram.items()]) / float((bcl - ps) * sum(re_histogram.values()))
                    else:
                        rber_calculated = -1.0
                    uber_calculated = sum([n_errs * count for n_errs, count in ue_histogram.items()]) / float(bl * sum(ue_histogram.values()))
                    raw_data = {'metadata' : metadata_no_p
                               , 'p' : [p]
                               , 'raw_data' : raw_histogram
                               , 'rber' : rber_calculated
                               , 'uber' : uber_calculated
                               , 're_data' : re_histogram
                               , 'ue_data': ue_histogram}
                    callback(run_uid, raw_data, *callback_args)

                if line_num == 1 or (line_num % 1000) == 0:
                    sys.stdout.write('\t[' + infilename + "] {0:.2f}".format(100.0 * f.tell() / float(file_size_bytes)) + '% complete  \r')
                line = f.readline()
        sys.stdout.write('\n')
